hazelcast/src/main/java/com/hazelcast/query/impl/CompositeValue.java

```diff
@@ -160,8 +160,11 @@ public boolean equals(Object o) {
         if (o == null || getClass() != o.getClass()) {
             return false;
         }
-
         CompositeValue that = (CompositeValue) o;
+
+        // We are not using Comparables.equal here since an externally
```

Comment:
Still, is it ok that we are using `Comparables.compare` in `compareTo` but not `equals`/`hashCode`?

Reply:
Yes, it _should_ be ok :) `compareTo` and `equals'/`hashCode` are serving for slightly different purposes: `compareTo` is used in ordered indexes where values are non-canonical (thus they must be compared using `Comparables`) and `equals'/`hashCode` is for unordered ones where values are canonical (thus they can be "compared" using the regular `equals`/`hashCode`, but we must ensure that every pair of values we "compare" are in a canonical form). Composite values never participate in any predicates other than indexed `CompositeEqualPredicate` and `CompositeRangePredicate`, so any query involving composite values is executed using indexes. `Comparables.equal` exists mostly for full scan queries, so we may just check the values for equality without converting them into a canonical form.

In theory, `compareTo` and `equals'/`hashCode` should never be mixed, since order and equality might be established in a different way in Java, if `compareTo` returns `0` that doesn't mean `equals` must return `true` and vice versa. But we mix them: `EqualPredicate` vs `GreaterLessPredicate`,  `EqualPredicate` vs `Comparison.NOT_EQUAL`.